include/IECoreScene/MeshAlgo.inl

```diff
@@ -0,0 +1,290 @@
+#define IECORESCENE_MESHALGO_INL
+
+namespace IECoreScene
+{
+
+namespace MeshAlgo
+{
+
+namespace Detail
+{
+
+template< typename T >
+void initializeMeshSplitter( const MeshPrimitive *mesh, const PrimitiveVariable &primitiveVariable, std::unordered_map< T, int > &idMap, const IECore::Canceller *canceller, std::vector< int > &meshIndices, std::vector< int > &faceRemap, std::vector< int > &faceIndices, bool mustFillIdMap = false )
+{
+	typedef IECore::TypedData<std::vector<T>> VarDataType;
+	const VarDataType *typeVarData = IECore::runTimeCast<const VarDataType>( primitiveVariable.data.get() );
+	if( !typeVarData )
+	{
+		throw IECore::Exception( "Attempting to split mesh using map with " + std::string( IECore::TypedData<T>::staticTypeName() ) + " keys, but given primitive variable is of type " + std::string( primitiveVariable.data->typeName() ) );
+	}
+
+	if( primitiveVariable.interpolation != IECoreScene::PrimitiveVariable::Interpolation::Uniform )
+	{
+		throw IECore::Exception( "Primitive variable passed to MeshSplitter must be uniform." );
+	}
+
+	if( !mesh->isPrimitiveVariableValid( primitiveVariable ) )
+	{
+		throw IECore::Exception( "Primitive variable passed to MeshSplitter must be valid." );
+	}
+
+	if( mesh->numFaces() == 0 )
+	{
+		// If we don't initialize anything, numMeshes() will return 0, meaning there is no valid context to
+		// call splitMesh in, which is correct for an empty mesh
+		return;
+	}
+
+	const std::vector<T> &data = typeVarData->readable();
+
+	// The first step of splitting requires us to count the number of output meshes, and assign every
+	// face an integer from 0 to numOutputMeshes - 1 for which output mesh it belongs to.  To allow for
+	// efficiency with different possible kinds of inputs, this can be done either by allocating the
+	// indicesBuffer with indices that can be used directly, or by setting indicesPointer to a list of
+	// indices that already work, and optionally setting remapIndices with a remapping from the original
+	// indices to indices that exactly fulfill our criteria.
+
+	int numOutputMeshes = 0;
+	std::vector<int> indicesBuffer;
+	const std::vector<int> *indicesPointer = nullptr;
+	std::vector<int> remapIndices;
+
+	// remapMin specifies the lowest value in the indicesBuffer that we need to remap - it shifts all accesses
+	// to the remapIndices, allowing remapIndices to be used when the lowest element is not 0
+	int remapMin = 0;
+
+	if( !primitiveVariable.indices )
+	{
+		if constexpr( std::is_same< T, int >::value )
+		{
+			// Special case for integer primVar.  An integer primvar is common, because it is produced by
+			// sources like the Gaffer MeshSegments node, and in this case, we can go much faster using
+			// a vector than a hash map.  This requires us to check the range of the data first though.
+
+			int dataMin = data[0];
+			int dataMax = data[0];
+			for( int i : data )
+			{
+				dataMin = std::min( i, dataMin );
+				dataMax = std::max( i, dataMax );
+			}
```

Comment:
As discussed, we won't bother with this here, since minmax_element performs badly on int vectors where comparisons are much cheaper than missed branch predictions.